### PR TITLE
OpenAI: /v1/embeddings compatibility

### LINKED ISSUE #2416: `/v1/embeddings` OpenAI compatible API endpoint
Your blog post mentions you're considering it. We'd love it so that we can point our RAG apps at ollama. Thanks!

### PR DESCRIPTION
**Pull request by royjhan**
In anticipation of #5127 
Resolves #2416 

```
curl http://localhost:11434/v1/embeddings \
  -H "Content-Type: application/json" \
  -d '{
    "input": ["hello", "hi"],
    "model": "all-minilm"
  }'
```

```
{
  "object": "list",
  "data": [
    {
      "object": "embedding",
      "embedding": [
        ...
      ],
      "index": 0
    },
    {
      "object": "embedding",
      "embedding": [
        ...
      ],
     "index": 1
    }
  ],
  "model": "all-minilm"
}
```

### CODE REVIEW

File: openai/openai.go [hunk 1]
```diff
@@ -93,11 +98,24 @@ type Model struct {
...
 type ListCompletion struct {
 	Object string  `json:"object"`
 	Data   []Model `json:"data"`
 }
 
+type EmbeddingList struct {
+	Object string      `json:"object"`
+	Data   []Embedding `json:"data"`
+	Model  string      `json:"model"`
+	// Usage  Usage       `json:"usage,omitempty"`
```
Comment (jmorganca) [inline openai/openai.go:116]: should we remove this?

Comment (royjhan) [reply openai/openai.go:116]: openai returns prompt tokens and total tokens but didn't seem like api/embeddings supports that, can remove it as long as we're cool with just omitting it

Comment (jmorganca) [reply openai/openai.go:116]: Let's leave it out for now and document that accordingly. We can always add it to `/api/embed` later and pass it through here

File: openai/openai.go [hunk 2]
```diff
@@ -262,6 +279,27 @@ func toListCompletion(r api.ListResponse) ListCompletion {
 	}
 }
 
+func toEmbedCompletion(model string, r api.EmbedResponse) EmbeddingList {
```
Comment (jmorganca) [inline openai/openai.go:282]: Is this the right name for this function? `toEmbeddingList` ?

File: openai/openai.go [hunk 3]
```diff
@@ -693,6 +763,33 @@ func CompletionsMiddleware() gin.HandlerFunc {
 			id:         fmt.Sprintf("cmpl-%d", rand.Intn(999)),
 		}
 
+		c.Writer = w
+		c.Next()
+	}
+}
+
+func EmbedMiddleware() gin.HandlerFunc {
```
Comment (jmorganca) [inline openai/openai.go:771]:
```suggestion
func EmbeddingsMiddleware() gin.HandlerFunc {
```

**Review by jmorganca — APPROVED** [review]

**Pull request merged**